haystack/pipeline_utils/rag.py

```diff
@@ -36,15 +39,107 @@ def build_rag_pipeline(
     :param prompt_template: The template to use for the prompt. If not given, a default template is used.
     :param embedding_model: The name of the model to use for embedding. If not given, BM25 is used.
     :param llm_api_key: The API key to use for the OpenAI Language Model. If not given, the value of the
+    :param retriever_class: The retriever class to use. If not given, it will be inferred from the document store.
     llm_api_key will be attempted to be read from the environment variable OPENAI_API_KEY.
     """
-    return _RAGPipeline(
-        document_store=document_store,
-        generation_model=generation_model,
-        prompt_template=prompt_template,
-        embedding_model=embedding_model,
-        llm_api_key=llm_api_key,
+    # create the retriever
+    retriever_clazz = resolve_retriever(document_store, retriever_class)
+    retriever = retriever_clazz(document_store=document_store)
+
+    # embedding model must be given, only sentence-transformers models are supported (for now)
+    # create the embedder
+    if not embedding_model:
+        raise ValueError("Embedding model must be given. Currently, only sentence-transformers models are supported.")
+    embedder = SentenceTransformersTextEmbedder(model_name_or_path=embedding_model)
+
+    # create the generator
+    generator = resolve_generator(generation_model, llm_api_key)
+
+    # create the internal pipeline instance
+    # will be deprecated in the near future
+    return _RAGPipeline(retriever=retriever, embedder=embedder, generator=generator, prompt_template=prompt_template)
```

Comment:
Would it be possible to move the easy to read code from this class to the function as well?

I am not sure, because this class also solves the input resolution in the run() method. If this isnt handled elsewhere, we still need the class.

Reason: The code (pipeline construction and connections) in the _RAGPipeline.init is super well-readable. If people can see this simple code already in the build_rag_pipeline function, it will help them adjust the pipeline much quicker.

Reply:
The input slot resolution isnt handled as expected. I reopened https://github.com/deepset-ai/haystack/issues/6101 and left the _RAGPipeline class for now as is